### ISSUE TITLE
Use HTTP Proxy

### ISSUE THREAD

**Issue by dsmaugy** [ansible/tower-operator#215]
How do I configure AWX to use a HTTP proxy? With the old installation method I recall it was simply setting a couple variables in the inventory file but I'm not sure how to do it through the operator.

**Comment by tchellomello** [comment]
@dsmaugy assuming that you want to export the variable to your `AWX` kind instance, you could use the options: `tower_web_extra_env` and  `tower_task_extra_env` to specify the proxy.  For example, take a look at this https://github.com/ansible/awx-operator/blob/devel/roles/installer/defaults/main.yml#L112-L120

```yaml
# Add extra environment variables to the AWX task/web containers. Specify as
# literal block. E.g.:
# tower_task_extra_env: |
#   - name: FOO
#     value: bar
#   - name: BAZ
#     value: bing
```

If you want to extend the documentation with a PR it will be very welcome!

**Comment by dsmaugy** [comment]
@tchellomello 

Thanks for your response! I put this in my `awx.yml` config:
```
  tower_task_extra_env: |
    - name: HTTP_PROXY
      value: <PROXY URL>
    - name: HTTPS_PROXY
      value: <PROXY URL>
    - name: NO_PROXY
      value: <NO PROXY>
  tower_web_extra_env: |
    - name: HTTP_PROXY
      value: <PROXY URL>
    - name: HTTPS_PROXY
      value: <PROXY URL>
    - name: NO_PROXY
      value: <NO PROXY>
```

But when I restarted the pods, the environment variable still doesn't seem to be set. 
```
kubectl exec pod/awx-b5f6cf4d4-kf6c5 -c awx-web -it -- /bin/bash
bash-4.4$ echo $HTTP_PROXY

bash-4.4$ echo $HTTPS_PROXY

bash-4.4$ echo $NO_PROXY

bash-4.4$
```
(same results for awx-task pod)

Am I forgetting anything else?

**Comment by tchellomello** [comment]
@dsmaugy once you updated the `awx` spec, the operator will do a reconciliation to apply the difference and because there is a new env, k8s should rollout a new replica for you. If you just killed the pod, you will ended up with the same status.

So to troubleshoot it, do this:

1. Double check if you applied `awx` spec has the expected configuration `tower_task_extra_env` and `tower_web_extra_env`
```sh
kubectl get awx <NAME> -o yaml 
```
2. If was already there, then check if the operator applied the changes on the deployment: 
```sh
kubectl get deployment <NAME> -o yaml | grep -i  _PROXY
```
3. If its there and you still don't see doing the `kubectl exec`, rollout a new pod
```sh
kubectl rollout restart deployment/<NAME>
```

**Comment by dsmaugy** [comment]
@tchellomello  Looks like I'm having trouble at Step 2. The environment variables do not get applied to the deployment.

If I take a look at operator logs, this is what I see:

```
PLAY RECAP *********************************************************************
localhost                  : ok=27   changed=0    unreachable=0    failed=1    skipped=25   rescued=0    ignored=0


-------------------------------------------------------------------------------
{
   "level":"error",
   "ts":1618518674.674209,
   "logger":"controller-runtime.controller",
   "msg":"Reconciler error",
   "controller":"awx-controller",
   "request":"awx/awx",
   "error":"event runner on failed",
   "stacktrace":"github.com/go-logr/zapr.(*zapLogger).Error\n\tpkg/mod/github.com/go-logr/zapr@v0.1.1/zapr.go:128\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler\n\tpkg/mod/sigs.k8s.io/controller-runtime@v0.6.0/pkg/internal/controller/controller.go:258\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\tpkg/mod/sigs.k8s.io/controller-runtime@v0.6.0/pkg/internal/controller/controller.go:232\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).worker\n\tpkg/mod/sigs.k8s.io/controller-runtime@v0.6.0/pkg/internal/controller/controller.go:211\nk8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1\n\tpkg/mod/k8s.io/apimachinery@v0.18.2/pkg/util/wait/wait.go:155\nk8s.io/apimachinery/pkg/util/wait.BackoffUntil\n\tpkg/mod/k8s.io/apimachinery@v0.18.2/pkg/util/wait/wait.go:156\nk8s.io/apimachinery/pkg/util/wait.JitterUntil\n\tpkg/mod/k8s.io/apimachinery@v0.18.2/pkg/util/wait/wait.go:133\nk8s.io/apimachinery/pkg/util/wait.Until\n\tpkg/mod/k8s.io/apimachinery@v0.18.2/pkg/util/wait/wait.go:90"
}
```

I tried this both on the 0.8.0 operator release and the most recent release pushed to dev.

**Comment by tchellomello** [comment]
@dsmaugy  are you running a `managed` PostgreSQL? Which was that one?

**Comment by dsmaugy** [comment]
@tchellomello Yep, this is the only custom postgres setting I have defined: `tower_postgres_storage_class: hpe-iscsi`

**Comment by tchellomello** [comment]
@dsmaugy I was able to reproduce a similar issue on the ` awx-operator`  side, however, after a few reconciliation loops, it worked fine for me. I'm currently working on patch.  If you check your deployment, you don't see the vars despite the error in the `awx-deploy` ? (`kubectl get deployments <NAME>`

**Comment by dsmaugy** [comment]
@tchellomello Nope, even after a `kubectl rollout restart` the deployment remains untouched

**Comment by tchellomello** [comment]
Alright, so to reproduce the issue we can do the following:
1. Install an `AWX` instance from scratch using the `devel`  image using a `managed`  DB. You will have something like below:

```yaml
NAMESPACE   NAME                        READY   STATUS    RESTARTS   AGE
default     changed-55b6994fc4-4pwkf    4/4     Running   0          7m38s
default     changed-postgres-0          1/1     Running   0          7m45s
```

2. Once the instance gets deployed, modify the `AWX`  kind and add the following:
```yaml
spec:
  tower_task_extra_env: |
       - name: CUSTOM_FOO
         value: foo 
       - name: CUSTOM_BAR
         value: bar 
  ```

3. Because  of the change in the `AWX`, the `awx-operator`  will apply the changes (rendering the new jinja2 template with the custom vars added) , however, the `awx-operator`  will fail with the error below:

```yaml
[awx-operator-b467bb4cd-c4b27] fatal: [localhost]: FAILED! => {"changed": false, "msg": "Failed to execute on pod changed-55b6994fc4-4pwkf due to : (0)\nReason: Handshake status 500 Internal Server Error\n"} 
PLAY RECAP *********************************************************************\r\nlocalhost                  : ok=31   changed=1    unreachable=0    failed=1    skipped=25   rescued=0    ignored=0   \r\n\n","job":"5440260128355399364","name":"upgrade-test-2","namespace":"default","error":"exit status 2","stacktrace":"github.com/go-logr/zapr.(*zapLogger).Error\n\tpkg/mod/github.com/go-logr/zapr@v0.1.1/zapr.go:128\ngithub.com/operator-framework/operator-sdk/pkg/ansible/runner.(*runner).Run.func1\n\tsrc/github.com/operator-framework/operator-sdk/pkg/ansible/runner/runner.go:239"} 
```

**Explanation**

If we look below, we can see the process from rolling out a new deployment due to the envVar change. 

```yaml
NAMESPACE   NAME                        READY   STATUS    RESTARTS   AGE
default     changed-55b6994fc4-4pwkf    4/4     Running   0          7m38s
default     changed-postgres-0          1/1     Running   0          7m45s

  ^ all running well then the change on the awx happens and the deployment gets updated 

default     changed-555b588cc9-f5f64    0/4     Pending   0          0s
default     changed-555b588cc9-f5f64    0/4     Pending   0          0s
default     changed-555b588cc9-f5f64    0/4     ContainerCreating   0          0s   <<<--- starting a new updated pod
default     changed-555b588cc9-f5f64    4/4     Running             0          4s

  ^  the new deployment rolls out and terminates the old

default     changed-55b6994fc4-4pwkf    4/4     Terminating         0          9m30s
default     changed-55b6994fc4-4pwkf    0/4     Terminating         0          9m31s
```

4. Now looking in detail on the playbook output we can see the `tower_pod_name`  used is the old one that has been already terminated, therefore, making the remaining steps in the playbook fail. 

```yaml

  ** applying the changes due to the variable change

awx-operator-b467bb4cd-c4b27] TASK [installer : Apply Resources] *********************************************
[awx-operator-b467bb4cd-c4b27] task path: /opt/ansible/roles/installer/tasks/main.yml:53 

  **then the next step is to get the pod information

 awx-operator-b467bb4cd-c4b27] TASK [installer : Get the resource pod information.] ***************************
[awx-operator-b467bb4cd-c4b27] task path: /opt/ansible/roles/installer/tasks/main.yml:66 

 *** now we set it to a variable, however, look the old pod which will terminate gets the place
 
awx-operator-b467bb4cd-c4b27] TASK [installer : Set the resource pod name as a variable.] ********************
[awx-operator-b467bb4cd-c4b27] task path: /opt/ansible/roles/installer/tasks/main.yml:79 
[awx-operator-b467bb4cd-c4b27] ok: [localhost] => {"ansible_facts": {"tower_pod_name": "changed-55b6994fc4-4pwkf"}, "changed": false} 

 awx-operator-b467bb4cd-c4b27] TASK [installer : Verify the resource pod name is populated.] ******************
[awx-operator-b467bb4cd-c4b27] task path: /opt/ansible/roles/installer/tasks/main.yml:83 
 awx-operator-b467bb4cd-c4b27] ok: [localhost] => {
 awx-operator-b467bb4cd-c4b27]     "changed": false,
 awx-operator-b467bb4cd-c4b27]     "msg": "All assertions passed"
[awx-operator-b467bb4cd-c4b27] } 
 awx-operator-b467bb4cd-c4b27] 
 awx-operator-b467bb4cd-c4b27] TASK [installer : Check for pending migrations] ********************************
[awx-operator-b467bb4cd-c4b27] task path: /opt/ansible/roles/installer/tasks/main.yml:88 
[awx-operator-b467bb4cd-c4b27] fatal: [localhost]: FAILED! => {"changed": false, "msg": "Failed to execute on pod changed-55b6994fc4-4pwkf due to : (0)\nReason: Handshake status 500 Internal Server Error\n"} 
 awx-operator-b467bb4cd-c4b27] 
 awx-operator-b467bb4cd-c4b27] PLAY RECAP *********************************************************************
 awx-operator-b467bb4cd-c4b27] localhost                  : ok=31   changed=1    unreachable=0    failed=1    skipped=25   rescued=0    ignored=0   
```

  As we can see, the pod name used was `changed-55b6994fc4-4pwkf `  which already has been terminated.

**Comment by dsmaugy** [comment]
Hey @tchellomello , thanks for looking at this. I tried the newest operator image and version but it still didn't fix it.

It turns out the problem was my yaml. My NO_PROXY variable starts off as a wildcard, so it looked something like this:
 `NO_PROXY: *.example.com` . The `*` at the beginning was the culprit the whole time 🤦 
 
 All I had to do was quote my variable and everything worked fine. Just writing this down in case anyone else makes the same mistake. Thanks again for helping out and fixing the other bug.
